# Create sample GVCFs

Create sample GVCF files for a father/mother/child trio.

# Initialize variables

```commandline
READS_DIR=/data/reads/human
REF_DIR=/data/reference/human
VAR_DIR=/data/variation/human
OUTPUT_DIR=/data/output

mkdir -p $READS_DIR
mkdir -p $REF_DIR
mkdir -p $VAR_DIR
mkdir -p OUTPUT_DIR

READS_FILE=$READS_DIR/HG00096.chrom20.ILLUMINA.bwa.GBR.low_coverage.20101123.bam
REF_FILE_GZ=$REF_DIR/human_g1k_v37.fa.gz
REF_FILE=$REF_DIR/human_g1k_v37.fa
VAR_INDELS_B37_GZ=$VAR_DIR/1000G_phase1.indels.b37.vcf.gz
VAR_DBSNP_B37_GZ=$VAR_DIR/dbsnp_138.b37.vcf.gz

VAR_INDELS_B37=$VAR_DIR/1000G_phase1.indels.b37.vcf
VAR_DBSNP_B37=$VAR_DIR/dbsnp_138.b37.vcf

OUTPUT_VCF=$OUTPUT_DIR/HG00096.g.vcf

ONLY_PAIRED_READS_FILE=$READS_DIR/HG00096.chrom20.ILLUMINA.bwa.GBR.low_coverage.20101123.paired.bam
SORTED_READS_FILE=$READS_DIR/HG00096.chrom20.ILLUMINA.bwa.GBR.low_coverage.20101123.paired.sorted.bam
```

# Download Reads
```commandline
wget -O $READS_FILE ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/data/HG00096/alignment/HG00096.chrom20.ILLUMINA.bwa.GBR.low_coverage.20101123.bam
```

# Investigate reads

```commandline
samtools view -H $READS_FILE
samtools flagstat $READS_FILE
```

# Ensure even number of reads (paired end)

```commandline
samtools view -b -f 2 -F 524 $READS_FILE > $ONLY_PAIRED_READS_FILE
```

# Sort reads
```commandline
samtools sort -n $ONLY_PAIRED_READS_FILE -o $SORTED_READS_FILE --threads 12
```

# Download Reference
```commandline
wget -O $REF_FILE_GZ ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz
```

# Download Known sites

```commandline
wget -O $VAR_INDELS_B37_GZ ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/1000G_phase1.indels.b37.vcf.gz
wget -O $VAR_DBSNP_B37_GZ ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/dbsnp_138.b37.vcf.gz
```

# Decompress the vcf files

```commandline
gunzip $VAR_INDELS_B37_GZ
gunzip $VAR_DBSNP_B37_GZ
```

# Create index feature files

```commandline
gatk IndexFeatureFile --input $VAR_INDELS_B37
gatk IndexFeatureFile --input $VAR_DBSNP_B37
```

# Decompress the reference file

```commandline
gunzip $REF_FILE_GZ
```

# Create a FASTA sequence dictionary file from the reference genome

```commandline
gatk CreateSequenceDictionary --REFERENCE $REF_FILE
```

# Create index file for reference fasta genome

```commandline
samtools faidx $REF_FILE
```

# Perform BWA Mem Index Image Creation on the reference genome
See https://gatk.broadinstitute.org/hc/en-us/articles/360037054812-BwaMemIndexImageCreator

```commandline
gatk BwaMemIndexImageCreator -I $REF_FILE -O $REF_FILE.img
```

Note: this can take 2 hours

Example output:

```bash
[bwa_index] Pack forward-only FASTA... 16.33 sec
[bwa_index] Construct SA from BWT and Occ... 1080.78 sec
[M::bwa_idx_load_from_disk] read 0 ALT contigs
06:03:00.181 INFO  BwaMemIndexImageCreator - Shutting down engine
[May 17, 2022 6:03:00 AM GMT] org.broadinstitute.hellbender.tools.BwaMemIndexImageCreator done. Elapsed time: 127.04 minutes.
```


#  Run the Spark analysis pipeline

Note: This can take 2 hours
```commandline
gatk ReadsPipelineSpark \
    --input  $SORTED_READS_FILE \
    --reference $REF_FILE \
    --known-sites $VAR_INDELS_B37 \
    --known-sites $VAR_DBSNP_B37 \
    --output $OUTPUT_VCF \
    -align \
    --conf 'spark.driver.memory=6g' \
    --conf 'spark.executor.memory=10g' \
    --remove-all-duplicates true \
    --java-options "-Xmx48g" \
    --native-pair-hmm-threads 16 \
    --emit-ref-confidence GVCF
```

# Expected output

Note: this can take 2 hours
```bash
...
02:00:37.052 INFO  ReadsPipelineSpark - Shutting down engine
[May 18, 2022 2:00:37 AM GMT] org.broadinstitute.hellbender.tools.spark.pipelines.ReadsPipelineSpark done. Elapsed time: 156.66 minutes.
```

# For trio data (assumed in next step), repeat for the father/mother/child trio

### Father reads

```commandline
wget -O $SAVE_PATH ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00731/alignment/HG00731.chrom20.ILLUMINA.bwa.PUR.low_coverage.20130422.bam
```
### Mother reads

```commandline
wget -O $SAVE_PATH ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00732/alignment/HG00732.chrom20.ILLUMINA.bwa.PUR.low_coverage.20130422.bam
```

### Mother reads

```commandline
wget -O $SAVE_PATH ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/HG00733/alignment/HG00733.chrom20.ILLUMINA.bwa.PUR.low_coverage.20130415.bam
```
