# Recalibrate and filter variants

Fit/apply a variant quality score model and filter low quality variants

## Initialize variables

```commandline
EXPERIMENT_NAME=trio_731_732_733

REF_DIR=/data/reference/human
REF=$REF_DIR/human_g1k_v37.fa
GDB_WORKSPACE=/data/genomics_db/$EXPERIMENT_NAME.gendb
GENOTYPE_GVCF=/data/genotype_gvcfs/$EXPERIMENT_NAME.g.vcf

KNOWN_SITES_DIR=/data/Genomics/known_sites/human

KNOWN_SITES_DIR_CONTAINER=/data/known_sites/human

RECALIBRATION_DIR=/data/Genomics/variant_recalibration
RECAL_FILE=$RECALIBRATION_DIR/$EXPERIMENT_NAME.recal
TRANCHES_FILE=$RECALIBRATION_DIR/$EXPERIMENT_NAME.tranches
RSCRIPT_FILE=$RECALIBRATION_DIR/$EXPERIMENT_NAME.plots.R
```

## Download resources

Resources are used for training/test sets for VQSR model.
```commandline
wget -O $KNOWN_SITES_DIR/hapmap_3.3.b37.vcf.gz ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/hapmap_3.3.b37.vcf.gz
wget -O $KNOWN_SITES_DIR/1000G_omni2.5.b37.vcf.gz ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/1000G_omni2.5.b37.vcf.gz
wget -O $KNOWN_SITES_DIR/1000G_phase1.snps.high_confidence.b37.vcf.gz ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/1000G_phase1.snps.high_confidence.b37.vcf.gz
wget -O $KNOWN_SITES_DIR/dbsnp_138.b37.vcf.gz ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/b37/dbsnp_138.b37.vcf.gz
```

## Unzip resources and create IndexFeatureFiles

```commandline
gunzip $KNOWN_SITES_DIR/hapmap_3.3.b37.vcf.gz
gunzip $KNOWN_SITES_DIR/1000G_omni2.5.b37.vcf.gz
gunzip $KNOWN_SITES_DIR/1000G_phase1.snps.high_confidence.b37.vcf.gz
```

```commandline
gatk IndexFeatureFile --input $KNOWN_SITES_DIR_CONTAINER/hapmap_3.3.b37.vcf
gatk IndexFeatureFile --input $KNOWN_SITES_DIR_CONTAINER/1000G_omni2.5.b37.vcf
gatk IndexFeatureFile --input $KNOWN_SITES_DIR_CONTAINER/1000G_phase1.snps.high_confidence.b37.vcf
gatk IndexFeatureFile --input $KNOWN_SITES_DIR_CONTAINER/dbsnp_138.b37.vcf
```

## Perform variant score recalibration

Fit/apply VQSR model to obtain quality scores for variants

```commandline
gatk VariantRecalibrator \
   -R $REF \
   -V $GENOTYPE_GVCF \
   --java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true' \
   --resource:hapmap,known=false,training=true,truth=true,prior=15.0 $KNOWN_SITES_DIR_CONTAINER/hapmap_3.3.b37.vcf \
   --resource:omni,known=false,training=true,truth=false,prior=12.0 $KNOWN_SITES_DIR_CONTAINER/1000G_omni2.5.b37.vcf \
   --resource:1000G,known=false,training=true,truth=false,prior=10.0 $KNOWN_SITES_DIR_CONTAINER/1000G_phase1.snps.high_confidence.b37.vcf \
   --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 $KNOWN_SITES_DIR_CONTAINER/dbsnp_138.b37.vcf \
   -mode SNP \
   -O $RECALIBRATION_DIR/$EXPERIMENT_NAME.recal \
   --tranches-file $RECALIBRATION_DIR/$EXPERIMENT_NAME.tranches \
   --rscript-file $RECALIBRATION_DIR/$EXPERIMENT_NAME.plots.R \
   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR
```

## Apply the VQSR model to filter low-quality variants

```commandline
gatk ApplyVQSR \
   -R $REF \
   -V $GENOTYPE_GVCF \
   -O $KNOWN_SITES_DIR_CONTAINER/$EXPERIMENT_NAME.recalibrated.vcf.gz \
   --tranches-file $TRANCHES_FILE \
   --recal-file $RECAL_FILE \
   -mode SNP
 
```

Optionally unzip the output vcf.gz file for downstream analysis
```commandline
gunzip $KNOWN_SITES_DIR_CONTAINER/$EXPERIMENT_NAME.recalibrated.vcf.gz
```
